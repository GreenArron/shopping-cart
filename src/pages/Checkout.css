.checkout-backdrop {
  position: fixed;
  bottom: 0;
  left: 0;
  width: 100%;
  height: 100vh;
  backdrop-filter: brightness(60%);
}

section.checkout {
  position: fixed;
  bottom: 0;
  left: 0;
  width: 100%;
  min-height: 28rem;
  max-height: 65vh;

  background-color: hsl(0, 0%, 90%);
  box-shadow: 1px 1px 5px 0px rgba(0, 0, 0, 0.35);
}

section.checkout .checkout-main {
  display: flex;
  flex-direction: row;
  align-items: start;
}

section.checkout .checkout-cart {
  display: flex;
  flex-wrap: nowrap;
  overflow-x: auto;
  width: 78%;
  gap: 1.5rem;
  min-height: 18rem;
  margin: 0 1rem;
  padding: 1rem 1rem;

  background-color: hsl(0, 0%, 80%);
  border-radius: 18px;

  scrollbar-color: hsl(0, 0%, 80%) white;
  scrollbar-width: thin;
}

.checkout-product {
  display: flex;
  flex-direction: column;
  align-items: center;
  flex-shrink: 0;
  user-select: none;

  margin-top: 4px;

  --product-card-width: 18rem;
  width: var(--product-card-width);
  height: auto;
  font-weight: 600;
  padding: 16px 8px;
  background-color: white;
  border-radius: 8px;
}

.checkout-product img {
  aspect-ratio: 1 / 1;
  max-width: 8rem;
  height: auto;
  object-fit: contain;
  pointer-events: none;
}

.checkout-product .product-info {
  margin-top: 28px;
  display: flex;
  flex-direction: column;
  gap: 4px;
}

.checkout-product .product-name {
  max-width: max(2rem, calc(var(--product-card-width) - 2rem));
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  user-select: contain;
}

.checkout-product .rating {
  position: relative;
  display: flex;
  justify-content: center;
  font-size: 0.75rem;
  padding-left: 4ch;
}

.checkout-product .price {
  background-color: #5eba5e;
  color: white;
  width: fit-content;
  align-self: center;
  padding: 6px;
  border-radius: 8px;
}

.checkout-product .buy-controls {
  width: 80%;
  display: flex;
  justify-content: end;
}

.checkout-decide {
  height: 100%;
  align-self: center;
  background-color: white;
  width: 18rem;
  margin: 1rem;
  padding: 1rem 1rem 1.5rem;
  border-radius: 16px;
  display: flex;
  flex-direction: column;
}

.checkout-decide ul.checkout-info {
  list-style: none;
}

.checkout-decide ul.checkout-info li {
  display: flex;
  justify-content: start;
  align-items: center;
  gap: 0.5rem;
}

.checkout-decide button {
  width: fit-content;
  padding: 1rem;
  color: var(--text);
  font-weight: 600;
}

.checkout-decide button.checkout {
  color: var(--text);
  font-family: Verdana, Geneva, Tahoma, sans-serif;
  background-color: hsl(50, 60%, 50%);
  margin: 0 auto;
}
