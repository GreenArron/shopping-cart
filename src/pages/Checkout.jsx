import "./Checkout.css";
import PropTypes from "prop-types";
import { useOutletContext } from "react-router-dom";
import SelectCount from "../components/SelectCount";
import Draggable from "../components/Draggable";
import _ from "lodash";
import { round } from "../utils/commons";

function CheckoutItem({ title, image, price, cartCount, onCountChange }) {
  return (
    <div className="checkout-product">
      <img src={image} />
      <div className="product-info">
        <div className="product-name">{title}</div>
        <div className="price">$ {price}</div>
      </div>
      <div className="buy-controls">
        <SelectCount count={cartCount} setCount={onCountChange} />
      </div>
    </div>
  );
}

CheckoutItem.propTypes = {
  id: PropTypes.number,
  title: PropTypes.string,
  image: PropTypes.string,
  price: PropTypes.number,
  category: PropTypes.string,
  cartCount: PropTypes.number,
  onCountChange: PropTypes.func,
  rating: PropTypes.shape({
    rate: PropTypes.number,
    count: PropTypes.number,
  }),
};

function Checkout() {
  // eslint-disable-next-line no-unused-vars
  const [cart, setCart, products] = useOutletContext();
  const boughtProducts = products
    .filter((product) => Object.keys(cart).includes(product.id.toString()))
    .map((product) => {
      return { ...product, cartCount: cart[product.id] };
    });
  const totalPrice = round(
    boughtProducts.reduce((currTotal, product) => currTotal + product.price, 0),
    2,
  );

  function changeCart(func, productId) {
    const currentCount = cart[productId] === undefined ? 0 : cart[productId];
    const newCount = Math.max(func(currentCount), 0);
    const changed = newCount - currentCount;

    if (changed !== 0) {
      if (newCount === 0) {
        setCart((currCart) => _.omit(currCart, productId));
      } else {
        setCart((currCart) => {
          return { ...currCart, [productId]: newCount };
        });
      }
    }
  }

  function confirmPurchase() {
    alert("You bought it! (imaginary redirect)");
  }

  return (
    <div className="checkout-backdrop">
      <section className="checkout">
        <h1>Your Cart:</h1>

        <div className="checkout-main">
          <Draggable rootClass="checkout-cart">
            {boughtProducts.map((product) => {
              return (
                <CheckoutItem
                  key={product.id}
                  {...product}
                  onCountChange={(func) => changeCart(func, product.id)}
                />
              );
            })}
          </Draggable>
          <div className="checkout-decide">
            <ul className="checkout-info">
              <li>
                <label htmlFor="checkout-total">Total: </label>
                <span id="checkout-total">{totalPrice}$</span>
              </li>
            </ul>
            <button className="checkout" onClick={confirmPurchase}>
              Confirm Purchase
            </button>
          </div>
        </div>
      </section>
    </div>
  );
}

export default Checkout;
