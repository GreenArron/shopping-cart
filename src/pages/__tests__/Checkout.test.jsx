import { useState } from "react";
import { it, expect } from "vitest";
import { MemoryRouter, Routes, Route, Outlet } from "react-router-dom";
import { render, waitFor } from "@testing-library/react";
import Checkout from "../Checkout";

const mockProducts = [
  {
    id: 1,
    title: "TestProduct",
    price: 451,
    description: "Perfect mock test product",
    category: "show-test",
    image: "",
    rating: {
      rate: 5,
      count: 10,
    },
  },
  {
    id: 2,
    title: "NestProduct",
    price: 100,
    description: "Second Product mocked",
    category: "show-test2",
    image: "",
    rating: {
      rate: 3,
      count: 2000,
    },
  },
];
function createMockApp(initialCart = {}) {
  function MockApp() {
    const [mockCart, setMockCart] = useState(initialCart);
    return (
      <div className="testing-mock">
        <Outlet context={[mockCart, setMockCart, mockProducts]} />
      </div>
    );
  }
  return <MockApp />;
}

async function renderCommon(cart = {}) {
  let renderedElement;

  await waitFor(() => {
    renderedElement = render(
      <MemoryRouter>
        <Routes location={"/products/checkout"}>
          <Route path="/" element={createMockApp(cart)}>
            <Route path="/products/checkout" element={<Checkout />} />
          </Route>
        </Routes>
      </MemoryRouter>,
    );
  });

  return renderedElement;
}

// these test: existence of buttons, total calculation, cart item showing up :P
// rest is tested on component level
it("matches last approved render (cart: 1:3)", async () => {
  const cart = { 1: 3 };
  const { container } = await renderCommon(cart);
  expect(container).toMatchSnapshot();
});

it("matches last approved render (cart: 1:1 2:1)", async () => {
  const cart = { 1: 1, 2: 1 };
  const { container } = await renderCommon(cart);
  expect(container).toMatchSnapshot();
});
