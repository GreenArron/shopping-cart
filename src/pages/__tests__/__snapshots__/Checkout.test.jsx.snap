// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches last approved render (cart: 1:1 2:1) 1`] = `
<div>
  <div
    class="testing-mock"
  >
    <div
      class="checkout-backdrop"
    >
      <section
        class="checkout"
      >
        <h1>
          Your Cart:
        </h1>
        <div
          class="checkout-main"
        >
          <div
            aria-label="Current Cart"
            class="checkout-cart"
          >
            <article
              class="checkout-product"
            >
              <img
                src=""
              />
              <div
                class="product-info"
              >
                <div
                  class="product-name"
                >
                  TestProduct
                </div>
                <div
                  class="price"
                >
                  $ 
                  451
                </div>
              </div>
              <div
                class="buy-controls"
              >
                <div
                  class="select-count"
                >
                  <button
                    aria-label="decrease"
                  >
                    <svg
                      viewBox="0 0 24 24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M19,4H15.5L14.5,3H9.5L8.5,4H5V6H19M6,19A2,2 0 0,0 8,21H16A2,2 0 0,0 18,19V7H6V19Z"
                      />
                    </svg>
                  </button>
                  <span>
                    1
                  </span>
                  <button
                    aria-label="increase"
                  >
                    +
                  </button>
                </div>
              </div>
            </article>
            <article
              class="checkout-product"
            >
              <img
                src=""
              />
              <div
                class="product-info"
              >
                <div
                  class="product-name"
                >
                  NestProduct
                </div>
                <div
                  class="price"
                >
                  $ 
                  100
                </div>
              </div>
              <div
                class="buy-controls"
              >
                <div
                  class="select-count"
                >
                  <button
                    aria-label="decrease"
                  >
                    <svg
                      viewBox="0 0 24 24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M19,4H15.5L14.5,3H9.5L8.5,4H5V6H19M6,19A2,2 0 0,0 8,21H16A2,2 0 0,0 18,19V7H6V19Z"
                      />
                    </svg>
                  </button>
                  <span>
                    1
                  </span>
                  <button
                    aria-label="increase"
                  >
                    +
                  </button>
                </div>
              </div>
            </article>
          </div>
          <div
            class="checkout-decide"
          >
            <ul
              class="checkout-info"
            >
              <li>
                <label
                  for="checkout-total"
                >
                  Total: 
                </label>
                <span
                  id="checkout-total"
                >
                  551
                  $
                </span>
              </li>
            </ul>
            <button
              class="checkout"
            >
              Purchase
            </button>
            <button
              class="continue"
            >
              Continue Shopping
            </button>
          </div>
        </div>
      </section>
    </div>
  </div>
</div>
`;

exports[`matches last approved render (cart: 1:3) 1`] = `
<div>
  <div
    class="testing-mock"
  >
    <div
      class="checkout-backdrop"
    >
      <section
        class="checkout"
      >
        <h1>
          Your Cart:
        </h1>
        <div
          class="checkout-main"
        >
          <div
            aria-label="Current Cart"
            class="checkout-cart"
          >
            <article
              class="checkout-product"
            >
              <img
                src=""
              />
              <div
                class="product-info"
              >
                <div
                  class="product-name"
                >
                  TestProduct
                </div>
                <div
                  class="price"
                >
                  $ 
                  451
                </div>
              </div>
              <div
                class="buy-controls"
              >
                <div
                  class="select-count"
                >
                  <button
                    aria-label="decrease"
                  >
                    -
                  </button>
                  <span>
                    3
                  </span>
                  <button
                    aria-label="increase"
                  >
                    +
                  </button>
                </div>
              </div>
            </article>
          </div>
          <div
            class="checkout-decide"
          >
            <ul
              class="checkout-info"
            >
              <li>
                <label
                  for="checkout-total"
                >
                  Total: 
                </label>
                <span
                  id="checkout-total"
                >
                  451
                  $
                </span>
              </li>
            </ul>
            <button
              class="checkout"
            >
              Purchase
            </button>
            <button
              class="continue"
            >
              Continue Shopping
            </button>
          </div>
        </div>
      </section>
    </div>
  </div>
</div>
`;
